### PR TITLE
Implement `fastlane_require` to automatically install gems into bundle

### PR DESCRIPTION
**Pull request by KrauseFx**
screenshot 2016-11-03 18 45 20

### CODE REVIEW

File: fastlane/lib/fastlane/fast_file.rb [hunk 1]
```diff
@@ -24,6 +24,13 @@ def initialize(path = nil)
                 'you should turn off smart quotes in your editor of choice.'
       end
 
+      content.scan(/.*require (.*)/).each do |current|
+        current = current.first
+        next if current.include?(".") # these are local gems
+        next if current.include?("/")
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fast_file.rb]: This could be improved

Comment (ohwutup) [inline fastlane/lib/fastlane/fast_file.rb]: What is this looking for exactly?

File: fastlane/lib/fastlane/fast_file.rb [hunk 2]
```diff
@@ -24,6 +24,13 @@ def initialize(path = nil)
                 'you should turn off smart quotes in your editor of choice.'
       end
 
+      content.scan(/.*require (.*)/).each do |current|
```
Comment (ohwutup) [inline fastlane/lib/fastlane/fast_file.rb]: Are we looking for `.*require` because we want to check both `fastlane_require` and regular `require`?

**Review by ohayon — APPROVED** [review]
I think this looks really awesome! 💯

**Comment by hjanuschka** [comment]
@KrauseFx will this work  on Fastfiles imported with `import_from_git` too? (would be awesome!)

File: fastlane/lib/fastlane/fast_file.rb [hunk 3]
```diff
@@ -24,6 +24,14 @@ def initialize(path = nil)
                 'you should turn off smart quotes in your editor of choice.'
       end
 
+      content.scan(/(.*require) (.*)/).each do |current|
+        require_command = current.first
+        gem_name = current.last
+        next if require_command.include?("fastlane_require")
```
Comment (milch) [inline fastlane/lib/fastlane/fast_file.rb]: Use a regex that only matches require and not `fastlane_require`, like `^\s*(require) (.*)`

File: fastlane/lib/fastlane/fast_file.rb [hunk 4]
```diff
@@ -179,6 +187,10 @@ def desc_collection
       @desc_collection ||= []
     end
 
+    def fastlane_require(gem_name)
+      FastlaneRequire.install_gem(gem_name: gem_name, require_gem: true)
```
Comment (milch) [inline fastlane/lib/fastlane/fast_file.rb]: `install_gem` should probably be called `install_gem_if_needed`

File: fastlane_core/lib/fastlane_core/helper.rb [hunk 5]
```diff
@@ -39,6 +39,12 @@ def self.bundler?
       return false
     end
 
+    # Do we run from a bundled fastlane, which contains Ruby and OpenSSL?
+    # Usually this means the fastlane directory is ~/.fastlane/bin/
+    def self.contained_fastlane?
+      ENV["FASTLANE_IS_BUNDLED"].to_s.length > 0
```
Comment (milch) [inline fastlane_core/lib/fastlane_core/helper.rb]: Maybe do an explicit check for `== "true"` here or something, or things will break if someone sets `ENV["FASTLANE_IS_BUNDLED"]="false"` in the future

**Review by milch — CHANGES_REQUESTED** [review]

File: fastlane_core/lib/fastlane_core/helper.rb (hunk 5 above)
Comment (milch) [inline fastlane_core/lib/fastlane_core/helper.rb]: Note: I've left this comment to link to in the other PR, this should be removed if we split this into 2 PR's

**Comment by KrauseFx** [comment]
@milch Thanks for the feedback, I updated the PR 👍

File: fastlane/lib/fastlane/actions/cocoapods.rb [hunk 6]
```diff
@@ -14,7 +14,7 @@ def self.run(params)
           cmd << ["cd '#{podfile_folder}' &&"]
         end
 
-        cmd << ['bundle exec'] if File.exist?('Gemfile') && params[:use_bundle_exec]
+        cmd << ['bundle exec'] if params[:use_bundle_exec] && shell_out_should_use_bundle_exec?
```
Comment (milch) [inline fastlane/lib/fastlane/actions/cocoapods.rb]: Still not 100% satisfied with this. I'd go for 

```ruby
def self.shell_bundle_exec_prefix
    shell_out_should_use_bundle_exec ? "bundle exec" : ""
end
```

and then do 

```ruby
cmd << shell_bundle_exec_prefix if params[:use_bundle_exec]
```

everywhere (or something).

Comment (milch) [inline fastlane/lib/fastlane/actions/cocoapods.rb]: Although I'm willing to conceded that that would be the cherry  on top, more or less, so if the `contained_fastlane?` thing is fixed I'd be okay to merge.

Comment (KrauseFx) [inline fastlane/lib/fastlane/actions/cocoapods.rb]: Yeah, let's try to get the first version working and refactor this later on.

Comment (milch) [inline fastlane/lib/fastlane/actions/cocoapods.rb]: Makes sense

**Review by milch — APPROVED** [review]

**Pull request merged**